x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/local/InferIsNotNull.java

```diff
@@ -106,6 +107,6 @@ private boolean doResolve(Expression exp, AttributeMap<Expression> aliases, Set<
     }
 
     private static boolean skipExpression(Expression e) {
-        return e instanceof Coalesce;
+        return e instanceof Coalesce || e instanceof Case;
```

Comment:
Nit: in case any other change is required to the PR, maybe move the comment from the call site as function javadoc/comment.